File: Data Structures & Algorithms/duplicate-integer/submission-4.py
class Solution:
    def hasDuplicate(self, nums: List[int]) -> bool:
        seen = set()
        for num in nums:
            if num in seen:
                return True
            seen.add(num)
        return False
            
        
        """
        unordered_map<int, int> umap;

        for (int i = 0; i < nums.size(); i++) 
        {
            if (umap.find(nums[i]) != umap.end()) 
            {
                return true;
            }
            umap[nums[i]] = 1; 
        }
        
        return false;
        """

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-4.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) 
    {
        int l = 0, r = nums.size() - 1;

        while(l <= r)
        {
            int m = l + (r - l) / 2;

            if(nums[m] == target) return m;

            if(nums[m] >= nums[l])
            {
                if(target > nums[m] || target < nums[l])
                {
                    l = m + 1;
                }
                else
                {
                    r = m - 1;
                }
            }
            else
            {
                if(target < nums[m] || target > nums[r])
                {
                    r = m - 1;
                }
                else
                {
                    l = m + 1;
                }
            }
        }

        return -1;
    }
};

/*
0 1 2 3 4 5 6 7
1 2 3 4 5 6 7 0
2 3 4 5 6 7 0 1
3 4 5 6 7 0 1 2
4 5 6 7 0 1 2 3
5 6 7 0 1 2 3 4
6 7 0 1 2 3 4 5
7 0 1 2 3 4 5 6
*/